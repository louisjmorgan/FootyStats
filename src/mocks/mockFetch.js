import mockMatches from 'mocks/responses/mockMatches.json';
import mockMatch from 'mocks/responses/mockMatch.json';
import mockEvents from 'mocks/responses/mockEvents.json';

export default async function mockFetch(url) {
  switch (url) {
    case '/api/matches': {
      return {
        ok: true,
        status: 200,
        json: async () => (mockMatches),
      };
    }
    case '/api/matches/1485449': {
      return {
        ok: true,
        status: 200,
        json: async () => (mockMatch),
      };
    }
    case '/api/matches/1485449/events': {
      return {
        ok: true,
        status: 200,
        json: async () => (mockEvents),
      };
    }
    default: {
      throw new Error(`Unhandled request: ${url}`);
    }
  }
}
