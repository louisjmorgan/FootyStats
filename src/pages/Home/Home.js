/* eslint-disable no-underscore-dangle */
import { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import {
  Button, Container, Stack, Form,
} from 'react-bootstrap';
import { getMatches } from 'api';

export default function Home() {
  const [matches, setMatches] = useState([]);
  const [filteredMatches, setFilteredMatches] = useState([]);
  const [teams, setTeams] = useState([]);

  useEffect(() => {
    getMatches().then((res) => {
      setMatches(() => res.matches);
      setFilteredMatches(() => res.matches);
    });
  }, []);

  useEffect(() => {
    setTeams(() => matches.reduce((prev, match) => {
      if (prev.indexOf(match.home.name) === -1) {
        prev.push(match.home.name);
      }
      return prev;
    }, []));
  }, [matches]);

  function handleSelectTeam(team) {
    if (team === 'all') {
      setFilteredMatches(() => matches);
      return;
    }
    setFilteredMatches(() => matches.filter((match) => {
      if (match.home.name === team || match.away.name === team) return true;
      return false;
    }));
  }

  return (
    <Container fluid className="App">
      <h1 className="p-5 text-center text-secondary">Footy Stats</h1>

      <Container fluid className="match-buttons  min-vh-50 p-5">
        <Form.Label className="w-100 d-flex flex-column justify-content-center mx-auto m-5 text-center">
          Select Team:
          <Form.Select className="w-25 mx-auto" onChange={(e) => handleSelectTeam(e.target.value)}>
            <option value="all">All</option>
            {
              teams.sort().map((team) => (
                <option key={team} value={team}>{team}</option>
              ))
            }
          </Form.Select>
        </Form.Label>
        <Stack className="w-75 mx-auto d-flex flex-wrap justify-content-center" direction="horizontal" gap={2}>
          {filteredMatches.map((match) => (
            <Link key={match._id} to={`/matches/${match._id}`}>
              <Button variant="outline-primary" className="border-2">
                {match.home.name}
                {' '}
                {match.ftScore}
                {' '}
                {match.away.name}
              </Button>
            </Link>
          ))}
        </Stack>
      </Container>
    </Container>
  );
}
