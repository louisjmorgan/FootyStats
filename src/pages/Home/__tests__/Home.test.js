import {render, screen} from "@testing-library/react";
import Home  from '../Home';
import { BrowserRouter } from 'react-router-dom';
import mockFetch from 'mocks/mockFetch'

beforeAll(() => jest.spyOn(global, 'fetch'));
beforeEach(() => global.fetch.mockImplementation(mockFetch));

describe("Correctly display list of matches as links", () => {

    test("obtains matches from api and renders them", async () => {
        render(
            <BrowserRouter>
                <Home />
            </BrowserRouter>
        );
        const matchLinkElements = await screen.findAllByRole('link');
        expect(matchLinkElements.length).toBe(2);
    });

    test("renders match participants and score in the link text", async () => {
        render(
            <BrowserRouter>
                <Home />
            </BrowserRouter>
        );
        const matchLinkElement = await screen.findByText('Wolves 1 : 1 Tottenham');
        expect(matchLinkElement).toBeInTheDocument();
    });

    test("link points to correct path", async () => {
        render(
            <BrowserRouter>
                <Home />
            </BrowserRouter>
        );
        const matchLinkElements = await screen.findAllByRole('link');
        expect(matchLinkElements[0]).toHaveAttribute('href', '/matches/1485449');
    })
})