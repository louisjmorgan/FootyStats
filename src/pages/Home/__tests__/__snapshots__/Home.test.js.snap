// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot matches snapshot 1`] = `
<div>
  <div
    class="App container-fluid"
  >
    <h1
      class="p-5 text-center text-secondary"
    >
      Footy Stats
    </h1>
    <div
      class="match-buttons  min-vh-50 p-5 container-fluid"
    >
      <label
        class="w-100 d-flex flex-column justify-content-center mx-auto m-5 text-center form-label"
      >
        Select Team:
        <select
          class="w-25 mx-auto form-select"
        >
          <option
            value="all"
          >
            All
          </option>
          <option
            value="Crystal Palace"
          >
            Crystal Palace
          </option>
          <option
            value="Wolves"
          >
            Wolves
          </option>
        </select>
      </label>
      <div
        class="w-75 mx-auto d-flex flex-wrap justify-content-center hstack gap-2"
      >
        <a
          href="/matches/1485449"
        >
          <button
            class="border-2 btn btn-outline-primary"
            type="button"
          >
            Wolves
             
            1 : 1
             
            Tottenham
          </button>
        </a>
        <a
          href="/matches/1485340"
        >
          <button
            class="border-2 btn btn-outline-primary"
            type="button"
          >
            Crystal Palace
             
            0 : 0
             
            Fulham
          </button>
        </a>
      </div>
    </div>
  </div>
</div>
`;
